### ISSUE TITLE
`variant` does not name a *tuple* variant or *tuple* struct for data-less variant

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#35377]
``` rust
enum A {
    V {}
}

fn main(){ 
    match (A::V {}) {
        A::V => {}
    }
}
```

results in

```
error: `A::V` does not name a tuple variant or a tuple struct [--explain E0164]
 --> <anon>:7:9
7 |>         A::V => {}
  |>         ^^^^
```

instead of pointing user out that this variant is struct-like and they forgot to match on it properly.

**Comment by petrochenkov** [comment]
The "tuple" part is fixed on nightly, now the error is
```
error[E0533]: `A::V` does not name a unit variant, unit struct or a constant
 --> <anon>:7:9
  |
7 |         A::V => {}
  |         ^^^^
```

**Comment by estebank** [comment]
[Current output](https://play.rust-lang.org/?gist=522fd5c63034ec04e1685074f267e9ab&version=nightly&backtrace=0):

```
error[E0532]: expected unit struct/variant or constant, found struct variant `A::V`
 --> <anon>:7:9
  |
7 |         A::V => {}
  |         ^^^^
```

**Comment by petrochenkov** [comment]
Fixed in 1.16.0
```
rustc 1.16.0-beta.2 (bc15d5281 2017-02-16)
error[E0532]: expected unit struct/variant or constant, found struct variant `A::V`
 --> <anon>:7:9
  |
7 |         A::V => {}
  |         ^^^^ did you mean `A::V { /* fields */ }`?

error: aborting due to previous error
```
